cdap-common/src/main/java/io/cdap/cdap/internal/app/store/OperationRunDetail.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright © 2015-2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.store;
+
+import com.google.common.base.Objects;
+import com.google.gson.annotations.SerializedName;
+import io.cdap.cdap.proto.operationrun.OperationError;
+import io.cdap.cdap.proto.operationrun.OperationMeta;
+import io.cdap.cdap.proto.operationrun.OperationRun;
+import io.cdap.cdap.proto.operationrun.OperationRunStatus;
+import java.util.Arrays;
+import javax.annotation.Nullable;
+
+/**
+ * Store the meta information about program runs in CDAP. This class contains all information the
+ * system needs about a run, which includes information that should not be exposed to users.
+ * {@link io.cdap.cdap.proto.operationrun.OperationRun} contains fields that are exposed to users,
+ * so everything else like the user request and principal goes here
+ *
+ * @param <T> The type of the operation request
+ */
+public class OperationRunDetail<T> extends OperationRun {
+
+  // sourceid refers to the tms message id which has updated this run details
+  @SerializedName("sourceid")
+  @Nullable
+  private final byte[] sourceId;
+
+  @SerializedName("principal")
+  @Nullable
+  private final String principal;
+
+  @SerializedName("request")
+  private final T request;
+
+  protected OperationRunDetail(String runId, String type, OperationRunStatus status,
+      OperationMeta metadata, @Nullable OperationError error,
+      byte[] sourceId, @Nullable String principal, T request) {
+    super(runId, type, status, metadata, error);
+    this.sourceId = sourceId;
+    this.principal = principal;
+    this.request = request;
+  }
+
+  @Nullable
+  public byte[] getSourceId() {
+    return sourceId;
+  }
+
+  @Nullable
+  public String getPrincipal() {
+    return principal;
+  }
+
+  public T getRequest() {
+    return request;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    OperationRunDetail<T> that = (OperationRunDetail<T>) o;
+    return Objects.equal(this.getRunId(), that.getRunId())
```

Comment:
should only check variables in this class, then check the parent equals() method. Otherwise it is very easy to break if the parent adds a new field and we forget to update this method.